packages/chain-mon/src/faultproof-wd-mon/service.ts

```diff
@@ -0,0 +1,519 @@
+        this.state.portal.filters.WithdrawalProven(),
+        this.state.highestUncheckedBlockNumber,
+        toBlockNumber
+      )
+    } catch (err) {
+      // Log the issue so we can debug it.
+      this.logger.error(`got error when connecting to node`, {
+        error: err,
+        node: 'l1',
+        section: 'querying for WithdrawalProven events',
+      })
+
+      // Increment the metric so we can detect the issue.
+      this.metrics.nodeConnectionFailures.inc({
+        layer: 'l1',
+        section: 'querying for WithdrawalProven events',
+      })
+
+      // Sleep for a little to give intermittent errors a chance to recover.
+      return sleep(this.options.sleepTimeMs)
+    }
+
+    // Go over all the events and check if the withdrawal hash actually exists on L2.
+    for (const event of events) {
+      // Could consider using multicall here but this is efficient enough for now.
+      const hash = event.args.withdrawalHash
+      const disputeGamesData = await this.getWithdrawalDisputeGames(event)
+      for (const disputeGameData of disputeGamesData) {
+        const disputeGame = disputeGameData.disputeGame
+        const rootClaim = await disputeGame.rootClaim()
+        const l2BlockNumber = await disputeGame.l2BlockNumber()
+        const isValidRoot = await this.isValidOutputRoot(
+          rootClaim,
+          l2BlockNumber
+        )
+        if (isValidRoot) {
+          // Check if the withdrawal exists on L2.
+          const exists = await this.state.messenger.sentMessages(hash)
+          // Hopefully the withdrawal exists!
+          if (exists) {
+            // Unlike below we don't grab the timestamp here because it adds an unnecessary request.
+            this.logger.info(`valid withdrawal`, {
+              withdrawalHash: event.args.withdrawalHash,
+            })
+
+            // Bump the withdrawals metric so we can keep track.
+            this.metrics.withdrawalsValidated.inc()
+          } else {
+            this.state.invalidProofWithdrawals.push(disputeGameData)
+            // Grab and format the timestamp so it's clear how much time is left.
+            const block = await event.getBlock()
+            const ts = `${dateformat(
+              new Date(block.timestamp * 1000),
+              'mmmm dS, yyyy, h:MM:ss TT',
+              true
+            )} UTC`
+
+            // Uh oh!
+            this.logger.error(`withdrawalHash not seen on L2`, {
+              withdrawalHash: event.args.withdrawalHash,
+              provenAt: ts,
+            })
+
+            // Change to forgery state.
+            this.state.forgeryDetected = true
+            this.metrics.isDetectingForgeries.set(
+              Number(this.state.forgeryDetected)
+            )
+          }
+        } else {
+          this.state.invalidProposalWithdrawals.push(disputeGameData)
+          this.logger.info(`invalid proposal`, {
+            withdrawalHash: event.args.withdrawalHash,
```

Comment:
Probably also worth computing `provenAt` for this too. Can just compute once in the loop.